commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala

```diff
@@ -4,13 +4,54 @@ import java.time.temporal.{ChronoUnit, Temporal, TemporalUnit}
 import java.time.{Duration, LocalDateTime}
 import java.util.UUID
 
-import play.api.libs.json.{JsValue, Json, OFormat}
+import com.gu.notifications.events.utils.Percentiles
+import play.api.libs.json.{JsValue, Json, OFormat, Writes}
+import com.gu.notifications.events.utils.OWriteOps._
+
+case class TimingPercentiles(
+  `10th`: LocalDateTime,
+  `20th`: LocalDateTime,
+  `30th`: LocalDateTime,
+  `40th`: LocalDateTime,
+  `50th`: LocalDateTime,
+  `60th`: LocalDateTime,
+  `70th`: LocalDateTime,
+  `80th`: LocalDateTime,
+  `90th`: LocalDateTime,
+  `95th`: LocalDateTime,
+  `99th`: LocalDateTime
+)
+
+object TimingPercentiles {
+  implicit val jf = Json.format[TimingPercentiles]
+}
 
 case class EventAggregation(
   platformCounts: PlatformCount,
   providerCounts: ProviderCount,
   timing: Map[LocalDateTime, Int]
-)
+) {
+
+  def timingPercentiles: Option[TimingPercentiles] = {
+    val allTimings: Seq[LocalDateTime] = timing
+      .toSeq
+      .flatMap { case (t, count) =>
+        Seq.fill(count)(t)
```

Comment:
This is quite heavy memory-wise. @alexduf Can you think of a better way to calculate percentiles?

Reply:
Yes, I think we have count of the total notification received.
If we take the timings and sort it, then iterate through it, we can stop when we have the total number of notification so far > than the percentile you're calculating. Something like that (definitely not working scala, just to get the logic)

```scala
val mySortedTimings: Seq[(LocalDateTime, Int)] = timing.toSeq.sort
val `80thPercentileCount` = total * 0.8
val (`80thPercentileTime`, _) = mySortedTimings.foldLeft((notificationSentTime, 0)) {
  case ((_, totalSoFar), (time, count)) if totalSoFar < `80thPercentileCount` => (time, count + totalSoFar)
  case ((time, totalSoFar), _) if totalSoFar >= `80thPercentileCount` => (time, totalSoFar)
}
```